test/test.oauth2common.ts

```diff
@@ -0,0 +1,415 @@
+        `client_id=${encodeURIComponent(reqBodyAuth.clientId)}` +
+        `&client_secret=${encodeURIComponent(reqBodyAuth.clientSecret || '')}`;
+      expectedOptions.data = `key1=value1&key2=value2&${credentials}`;
+
+      handler.testApplyClientAuthenticationOptions(actualOptions);
+      assert.deepStrictEqual(expectedOptions, actualOptions);
+    });
+
+    it('should inject secretless creds in x-www-form-urlencoded content', () => {
+      const handler = new TestOAuthClientAuthHandler(reqBodyAuthNoSecret);
+      const originalOptions: GaxiosOptions = {
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        data: 'key1=value1&key2=value2',
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      const credentials =
+        `client_id=${encodeURIComponent(reqBodyAuthNoSecret.clientId)}` +
+        '&client_secret=';
+      expectedOptions.data = `key1=value1&key2=value2&${credentials}`;
+
+      handler.testApplyClientAuthenticationOptions(actualOptions);
+      assert.deepStrictEqual(expectedOptions, actualOptions);
+    });
+
+    it('should inject creds in empty x-www-form-urlencoded content', () => {
+      const handler = new TestOAuthClientAuthHandler(reqBodyAuth);
+      const originalOptions: GaxiosOptions = {
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      expectedOptions.data =
+        `client_id=${encodeURIComponent(reqBodyAuth.clientId)}` +
+        `&client_secret=${encodeURIComponent(reqBodyAuth.clientSecret || '')}`;
+
+      handler.testApplyClientAuthenticationOptions(actualOptions);
+      assert.deepStrictEqual(expectedOptions, actualOptions);
+    });
+  });
+
+  it('should process request with bearer token when provided', () => {
+    const bearerToken = 'BEARER_TOKEN';
+    const handler = new TestOAuthClientAuthHandler();
+    const originalOptions: GaxiosOptions = {
+      url: 'https://www.example.com/path/to/api',
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      data: {
+        key1: 'value1',
+        key2: 'value2',
+      },
+    };
+    const actualOptions = Object.assign({}, originalOptions);
+    const expectedOptions = Object.assign({}, originalOptions);
+    (expectedOptions.headers as Headers).Authorization = `Bearer ${bearerToken}`;
+
+    handler.testApplyClientAuthenticationOptions(actualOptions, bearerToken);
+    assert.deepStrictEqual(expectedOptions, actualOptions);
+  });
+
+  it('should prioritize bearer token over basic auth', () => {
```

Comment:
really happy with the tests that have been added.

Reply:
👍
Code coverage for this file:
```
File                                   | % Stmts | % Branch | % Funcs | % Lines |
  oauth2common.ts                      |     100 |    90.63 |     100 |     100 |
```